File: 212_space-replacement/space-replacement.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   ultimate010
@Problem:  http://www.lintcode.com/problem/space-replacement
@Language: Python
@Datetime: 16-06-15 12:23
'''

class Solution:
    # @param {char[]} string: An array of Char
    # @param {int} length: The true length of the string
    # @return {int} The true length of new string
    def replaceBlank(self, string, length):
        # Write your code here
        ret = ''
        spaceCount = 0
        for pos in range(length):
            if string[pos] == ' ':
                spaceCount += 1
        resultLen = spaceCount * 2 + length
        for pos in range(length - 1, -1, -1):
            if string[pos] == ' ':
                string[pos:pos+3] = '%20'
                spaceCount -= 1
                string[pos+spaceCount * 2:pos+spaceCount * 2 + 3] = string[pos:pos+3]
            else:
                string[pos + spaceCount * 2] = string[pos]
        return resultLen

File: 212_space-replacement/space-replacement.cpp
/*
@Copyright:LintCode
@Author:   ultimate010
@Problem:  http://www.lintcode.com/problem/space-replacement
@Language: C++
@Datetime: 16-08-04 01:38
*/

class Solution {
public:
    /**
     * @param string: An array of Char
     * @param length: The true length of the string
     * @return: The true length of new string
     */
    int replaceBlank(char string[], int length) {
        // Write your code here
        int spaceCount = 0;
        
        for (int i = 0; i < length; i++)
            if (string[i] == ' ')
                spaceCount++;
        int ret = length + spaceCount * 2;        
        for (int i = length - 1;i >= 0; i--){
            if (string[i] == ' '){
                spaceCount--;
                strncpy(string + i + spaceCount * 2, "%20", 3);
            } else {
                string[i + spaceCount * 2] = string[i];
            }
        }
        return ret;
    }
};  